rfcs/0077-layout-transform-padding.md

```diff
@@ -0,0 +1,2522 @@
+the padding value must be written to the padding prior to the
+completion of the operator.  Effectively, the producer must have a
+postlude as follows:
+
+```python
+for transformed_indices in T.grid(*transformed_shape):
+    if padding_predicate(*transformed_indices):
+        B[transformed_indices] = pad_value(*transformed_indices)
+```
+
+For a consumer of the transformed buffer, these padding values are
+initially unused, but may be used in later simplifications.
+
+## Overcompute vs Branching
+
+Depending on the computation being performed and the value stored in
+the padding, there can be trade-offs between branching and
+overcompute.  For example, consider the following `PrimFunc`, which
+computes the sum over each row of the input data.
+
+```python
+@T.prim_func
+def row_summation(a: T.handle, b: T.handle):
+    A = T.match_buffer(shape=(16, 14), dtype="float32")
+    B = T.match_buffer(shape=(16,), dtype="float32")
+    for i in T.serial(16):
+        B[i] = 0.0
+        for j in T.serial(14):
+            B[i] = B[i] + A[i, j]
+```
+
+We'd like to transform the layout of buffer `A` from `[i, j]` to `[i,
+j//4, j%4]`, along with the loop iteration.  By default, after using
+the `transform_layout` and `split` metaschedule primitives, we have
+the following function.
+
+```python
+@T.prim_func
+def row_summation(a: T.handle, b: T.handle):
+    A = T.match_buffer(shape=(16, 4, 4), dtype="float32")
+    B = T.match_buffer(shape=(16,), dtype="float32")
+    for i in T.serial(16):
+        B[i] = 0.0
+        for j_outer, j_inner in T.grid(4, 4):
+            if 4*j_outer + j_inner < 14:
+                B[i] = B[i] + A[i, j_outer, j_inner]
+```
+
+If the conditional can be removed, this function would be much more
+amenable for later vectorization, or to reduce branch divergence when
+bound to a thread index.  If the padding in `A` is pre-filled with
+zero, then `B[i] = B[i] + 0.0` is a no-op, and can be performed
+without changing the final computation.
+
+```python
+@T.prim_func
+def row_summation(a: T.handle, b: T.handle):
+    A = T.match_buffer(shape=(16, 4, 4), dtype="float32")
+    B = T.match_buffer(shape=(16,), dtype="float32")
+    for i in T.serial(16):
+        B[i] = 0.0
+        for j_outer, j_inner in T.grid(4, 4):
+            B[i] = B[i] + A[i, j_outer, j_inner]
+```
+
+By annotating the layout transformation with the value stored in the
+padding, this condition can be proven, allowing this conditional to
+automatically be removed.  Since the tradeoff between branching and
+overcompute may or may not be beneficial dependent on the schedule,
+these options are exposed as two additional transformations,
+`tir.transform.RemoveBranchingThroughOvercompute` and
+`tir.transform.RemoveOvercomputeThroughBranching`.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## TIR Changes
+
+### Buffer Annotation of Padding Predicate/Constraint Pairs
```

Comment:
Thanks @Lunderberg. 

It would be great to discuss ideas to simplify it further. While on one hand it is useful to introduce additional semantics to the IR itself. Doing so would generally affect the complexities around analysis and transformation.

It would be great for us to explore such capabilities without introducing additional complexities on the IR.  Back to our goal of introducing padding. It should be possible to have an explicit buffer transformation stage that coppies the data into the target padded buffer(with predication) , then run computation on the padded value.

There is certainly some tradeoffs here, but decoupling the padding behavior as a separate stage of IR computation should be able to allow us to reuse more primitives without having to specializing for BufferConstraint

Reply:
i agree it's helpful to think through alternatives here. could we consider some example transformations we may want to make (e.g. eliding or moving the operations which write to the padding) or pattern-matching on such operations and reducing them to hardware intrinsics (e.g. perhaps there is a way to tell the hardware how much padding to include when the value is always constant and a particular operation is in use). 

on the one hand, modeling the padding computation explicitly in TIR is a more logical reuse of existing TIR. on the other hand, it may be more expensive to match this and the compiler may be slower. 

i'm not necessarily in favor fo any one solution, but i think this is the sort of thing we should discuss to try and inform that decision.

Reply:
@tqchen @Hzfengsy Thank you, and I definitely agree on minimizing the number of IR changes being made.  (Also, phew, this ended up being a longer reply than I had expected, which probably means that whatever the result of this thread, the "Rationale and Alternatives" section should be updated.)

@areusch The example transformations are largely present in the "Implementation Options" section.  The goal of that section was to describe different example transformations that we'd like to be able to make, and to ensure that they could be made using the functionality introduced earlier in the RFC.  It wasn't until this morning that I realized that there should also be links in the other direction, pointing from the proposed IR changes to the motivating use case.

Below is the general rationale, with high-level implementations.

Starting by listing out the desired properties of an implementation.

1. No changes to existing TIR data structures
2. No additional meaning to existing TIR data structures
3. Simplifications can use constraints from multiple buffers
4. No ordering requiring between `transform_layout` and fuse/split/reorder.  (Conditional statements that are removed using the buffer constraints are typically introduced by loop rewrites.)
5. Can be used to describe out-of-bounds access (e.g. texture memory clamping on a GPU) that returns a default value.
6. Only allocate memory when required or requested


Implementations considered

- A. All buffer transformations introduce new stage

  - Pro: No coordination required between different operators.
  - Con: Any producer/consumer interactions must be recognized by operator fusion/device planning.
  - Con: Cannot apply to primfunc input/outputs.  (e.g. To de-duplicate operators that differ only by underlying layout, such as `topi.nn.conv2d_hwcn`, `topi.nn.conv2d_nchw`, `topi.nn.conv2d_NCHWc`, etc.)
  - Con: May introduce unnecessary data copies, if the constraint required by the consumer is already met.

- B. Perform one `transform_layout` at a time.  For each one, simplify using provided constraints, do not store constraints afterward.
  
  - Pro: Main downside is that it could only use the constraints of a single buffer at a time.  This wouldn't be able to express simplifications that rely on the padding in multiple buffers. (e.g. [elementwise operator](https://github.com/Lunderberg/tvm-rfcs/blob/buffer_layout_padding/rfcs/0077-layout-transform-padding.md#apply-operator-element-wise-over-the-transformation-padding)

  - Con: Requires loop rewriting to be done either inside `layout_transform` or prior to calling `layout_transform`.
  - Con: Can't be applied to use cases outside of layout transformations (e.g. texture memory clamping on a GPU), where simplifications could benefit from assumed constraints.
  
  
- C. Perform all `transform_layout` in a single function call, passing all layout transforms and padding constraints.

  - Pro: Simplifications may use constraints of all buffers being transformed.
  - Con: Requires changing the calling convention for layout transformations.
  - Con: Requires loop rewriting to be done either inside `layout_transform` or prior to calling `layout_transform`.
  - Con: Can't be applied to use cases outside of layout transformations (e.g. texture memory clamping on a GPU), where simplifications could benefit from assumed constraints.
    
- D. Express buffer constraints using existing `AssertStmt`

  In pseudocode, each consumer would have roughly the loopnest below. However, this would still need to have some way of indicating that the constraint should be removed when lowering, and should not produce any runtime assertions.
  
  ```python
  for indices in T.grid(*transform_shape):
      if padding_predicate(indices):
          T.Assert(buf[indices] == pad_value(indices))
  ```

  - Pro: No change to TIR data structures
  - Pro: No change required for calling convention for layout transformations.
  - Pro: Simplifications may use constraints of all buffers being transformed.
  - Pro: Can be applied to use cases outside of layout transformations (e.g. texture memory clamping on a GPU), where simplifications could benefit from assumed constraints.
  - Pro: No ordering between loop/layout transforms, because the constraints can be determined from the TIR.
  - Con: Additional meaning attached to existing TIR data structures.
  - Con: Can only describe a fixed number of assertions, wouldn't be able to express a default value for all out-of-bounds reads.

- E. Express buffer constraints as a field in `PrimFuncNode::attrs`

  - Con: Passes that replace `Buffer` objects must be aware of this attribute, in order to update the `Buffer` object stored in it.
    
- F. Express buffer constraints as a new member variable in `Buffer`

  - Con: Changes change to TIR data structures
  - Pro: No change required for calling convention for layout transformations.
  - Pro: Simplifications may use constraints of all buffers being transformed.
  - Pro: Can be applied to use cases outside of layout transformations (e.g. texture memory clamping on a GPU), where simplifications could benefit from assumed constraints.
  - Pro: Can rewrite loop structure later, use existing constraints.
  
  
1. No changes to existing TIR data structures
2. No additional meaning to existing TIR data structures
3. Simplifications can use constraints from multiple buffers
4. No ordering requiring between `transform_layout` and fuse/split/reorder.
5. Can be used to describe out-of-bounds access (e.g. texture memory clamping on a GPU) that returns a default value.
6. Only allocate memory when required or requested
  
- A. All buffer transformations introduce new stage
- B. Perform one `transform_layout` at a time.  For each one, simplify using provided constraints, do not store constraints afterward.
- C. Perform all `transform_layout` in a single function call, passing all layout transforms and padding constraints.
- D. Express buffer constraints using existing `AssertStmt`
- E. Express buffer constraints as a field in `PrimFuncNode::attrs`
- F. Express buffer constraints as a new member variable in `Buffer`
  
  
|        | Goal 1 | Goal 2 | Goal 3 | Goal 4 | Goal 5 | Goal 6 |
|--------|--------|--------|--------|--------|--------|--------|
| Impl A | :heavy_check_mark:    | :heavy_check_mark:    | :x:    | :x:    | :x:    | :x:    |
| Impl B | :heavy_check_mark:    | :heavy_check_mark:    | :x:    | :x:    | :x:    | :heavy_check_mark:    |
| Impl C | :heavy_check_mark:    | :heavy_check_mark:    | :heavy_check_mark:    | :x:    | :x:    | :heavy_check_mark:    |
| Impl D | :heavy_check_mark:    | :x:    | :heavy_check_mark:    | :heavy_check_mark:    | :x:    | :heavy_check_mark:    |
| Impl E | :heavy_check_mark:    | :x:    | :heavy_check_mark:    | :heavy_check_mark:    | :heavy_check_mark:    | :heavy_check_mark:    |
| Impl F | :x:    | :heavy_check_mark:    | :heavy_check_mark:    | :heavy_check_mark:    | :heavy_check_mark:    | :heavy_check_mark:    |

The implementations that would satisfy the largest number of the desired goals would be adding the member variable `BufferNode::constraints`, or adding a field to `PrimFuncNode::attrs` that holds the constraints.  Between the two, I lean toward having it as an explicit member variable, so that incorrect usage appears as a compilation error when compiling TVM, but would find either implementation acceptable.

Reply:
Thanks @Lunderberg  for dissected discussions, this is helpful.

Besides Goal 1, there is one additional implied goal:

- Goal 7: The compositionality of primitives of the buffer with layout-constraints. Specifically, how composable are the existing and new primitives(such as split/reorder/tensorization, reduction factorization) with the buffer layout constraints.

When building abstractions like this one, we are actually trying to make a balance among two things: the simplicity/composationality and the things we can support. It is quite natural that a more complicated impl would hit more marks initially.

On the other hand, there is always a consideration of added complexity. Additional field in the IR effectively means we either have to (a) introduce specific codepath to handle layout constraints, (b) generalize all relevant primitives to take that into account (introduces a N * M problem, where N is number of primitives and M is number of possible IR variantions(like layout constraints) we introduce to the IR). The N * M problem will grow as N and M increases.

It is also useful to come back to the high-level goal besides these goals for a single function. Our high-level goal is to enable effective end to end models under a good native layout(which involves padding and layout transformation). And it would actually be really nice to have an example at the e2e level to show how the set of transformations affect our optimizations.

Among the existing goals listed. Goal 6 is certainly a very important one. Goal 3 is primarily an implementation difference as in terms of different ways of building pattern matching. Goal 4 is not necessarily a need as many optimizations actually benefit from reduced complexity (e.g. tensorization in physical memory)

Goal 6 is an important one that would indeed touches the high-level (e2e) goal itself. Along that direction, a smart variant of Impl A (by interacting with graph) would actually enable simpler realization of goal 6 (which is important) by lifting the transformations of input/output out, and then cancel out between operators, while preserving the information.

Reply:
Thank you, and especially with the explicit Goal 7, that helps to clarify where I had been thinking about it differently.

> It is quite natural that a more complicated impl would hit more marks initially.  On the other hand, there is always a consideration of added complexity and how composable our additions are with existing constructs.

Definitely a good point.  (My goal in the RFC was to improve existing utilities wherever as possible, such as the changes to `StmtSimplifier` and `RemoveNoOp`, rather to introducing too many single-use passes.)  My concern was that attempting to describe a constraint in terms of existing TIR constructs and `AssertStmt` would be extremely fragile (e.g. loop over an input buffer and assert the constraint), and would lead other passes to need to be away of the composite structure in order to avoid breaking the structure.

On the composability side, based on yours and @Hzfensy's feedback, it sounds like the proposed scheduling primitives are one of the main issues, as they go against the typical design.  Since they are simplification-dependent, make multiple changes, and in some cases are composable from existing functionality, perhaps it would help to change them from stand-alone scheduling primitives into utilities that analyze a function and return a list of primitives that together have the desired effect.

> Along that direction, a smart variant of Impl A (by interacting with graph) would actually enable simpler realization of goal 6 (which is important) by lifting the transformations of input/output out, and then cancel out between operators, while preserving the information.

Enabling this interaction was my intent, so perhaps that means I have an error in how I'm picturing the relay/TIR interactions.  I had been picturing the `BufferConstraint` as the data structure in which to preserve the information and by which to passing the information between levels of abstraction.  When optimizing a single operator/PrimFunc, the layout and constraints of the arguments must remain constant, as otherwise it would be incompatible with neighboring operators.  The relay-level transforms would decide on buffer layout/constraints, and would specify those constraints to the TIR-level through `BufferConstraint`.

> And it would actually be really nice to have an example at the e2e level to show how the set of transformations affect our optimizations.

While not entirely end-to-end, I think the optimizations that would require coordination between multiple operators are briefly touched on in the example sections on [producer/consumer pairs](https://github.com/Lunderberg/tvm-rfcs/blob/buffer_layout_padding/rfcs/0077-layout-transform-padding.md#explicitly-write-next-operators-desired-default-at-end-of-function) and on [chainable `conv1d`](https://github.com/Lunderberg/tvm-rfcs/blob/buffer_layout_padding/rfcs/0077-layout-transform-padding.md#implicitly-write-default-value-of-next-operator).

Reply:
Thanks @Lunderberg for providing the details. To add a discussion point, I'd like to ask whether the semantic like over computation and writing default value of next operate, can be achieved with graph level rewriting. I agree that certain feedback loop between TIR and the Relay graph is needed. In Relay, this was previously achieved by some registration mechanism like (`AlterOpLayout`). We can use similar mechanism to guide the graph rewriting. For example, in chained elemwise operators, since we don't care about the undef value in the padding region, if we rewrite the graph to so that each operator has a shape that fits the layout requirement perfectly (which will directly generate the final TIR [here](https://github.com/Lunderberg/tvm-rfcs/blob/buffer_layout_padding/rfcs/0077-layout-transform-padding.md#apply-operator-element-wise-over-the-transformation-padding), would it be helpful to avoid the complexity in each individual operator involved dealing with the padding block and the conditions?

Reply:
Some notes from a conversation with @vinx13 @csullivan @tqchen @junrushao1994.  Wording is mine, attempting to summarize statements made.

* Hoist transformations into graph-level vs apply transformations in TIR.

  From @vinx13: The RFC proposal models layout transformations in TIR primarily focused on the TIR side.  However, this isn't strictly necessary, as padding could instead be introduced and reasoned about at the relay/relax level.

  Example: A single operator that performs a conv1d could be replaced by a sequence of three operators, to transform the layout, perform a conv1d, then apply the inverse transform.  If two adjacent conv1d use the same transform, then the transforms could be canceled out.

  From @Lunderberg: Uncertain how generic this could be for padding, as it would require a pre-existing implementation of the operator applied to a padded buffer, and would require graph-level knowledge of what padding can be applied to different buffers without effect.

  Buffer padding can also be trickier to reason about.  A layout transform of NCHW to NHWC followed by a layout transformation of NHWC to NCHW cancel each other out.  If a buffer is first cropped, then padded with zeros, these only cancel each other out if the contents of the cropped locations can be proven to have held zero.

* Optimization of PrimFuncs

  From @Lunderberg: Had been visualizing the layout transformations not solely as transformations that would be made in isolation, but as transformations that would need to be done in conjunction with the calling scope.

  When optimizing a PrimFunc in isolation, `transform_layout` would only be allowed to be applied on internal buffers, not buffers passed in as arguments.

* Graph-level optimization of buffer layouts.

  From @Lunderberg: Combining the previous two, this sounds like something that will be useful and doable in Relax, where a transformation could change both the input shape, and also the calling convention of all TIR functions that use the buffer.

  In relay, this would be trickier, but still possible.  Would need some way to query what layouts/assumptions an operator would find useful, either through manual tagging or some automatic search. Relay would first query the operators, then decide the layouts, then pass these down to the operators to use.

* Method to write compute definitions, especially for new hardware.

  From @csullivan: Modeling the layout transformation within TIR would have huge benefits when supporting new hardware targets.  When writing a new operator, there isn't a specific "graph-level" that it is being designed for.  The current state is quite difficult for hardware vendors, who must write a new compute definition and schedule for each shape of an operator.

  This is a benefit that wouldn't be present in the graph-level layout transformations, which would still require each supported layout to have a different implementation.

  From @Lunderberg: Writing a function that outputs a PrimFunc that follows a specific layout as a reasonable starting point for optimization would be a goal, and that writing in terms of a series of changes made to an initial PrimFunc would be the reasonable way to go about it.  Whether that is best described as a schedule primitive, a function that applies a schedule primitive, or a function that returns a series of schedule primitives to be applied is less important than the functionality itself.

Reply:
I think the general idea of reusability  A0 and A1 are not that different. Notably, A0 brings the reusablity in a sense of the utility simplifications wrt to folding etc. While A1 brings the simplifcations in the form of TIR level.

Note that compute defs in A0 can remain the same, the main difference is how the preferable schedules are being derived per informed by hw target as they are scheduled.

From a pure impl pov, considering the e2e goal. Having padding being sorted out in the graph level, actually still simplifies the scheduling layer. 

Note that A0 still does not preclude us from doing constraint matching, one can view that as insert a pre-proc stage that "must simplifies". At the high-level it is only a repr difference(of putting in the buffer decl vs stages).

Under certain scenarios we could indeed consider put some constraints at the interface level. That would need some more thoughts on semantics, how would they interact with graph, and structural complexity (whether padding is something that worth the IR complexity). 

For most e2e goals perhaps having a padding throughout graph level is not a bad way to reduce that part of the complexity.

Reply:
I was trying to work through what the transformation hoisting would look like at the graph level, and I think it runs into the same requirement of tracking the buffer constraints.  The derivation uses a toy model, consisting of the two sequential 1-d convolutions from [this section](https://github.com/Lunderberg/tvm-rfcs/blob/buffer_layout_padding/rfcs/0077-layout-transform-padding.md#implicitly-write-default-value-of-next-operator).  Layout transformations are introduced in read/write caches, and those layout transformations are hoisted into separate functions, so that they can be manipulated at the graph level.  The adjacent transformations are then fused and simplified.

(The full derivation is in [this gist](https://gist.github.com/Lunderberg/7dcd4edbdd7bedfb08072037792aa585), as including it here made the comment unreadably long.)

The key result from the derivation is that, for a padded transformation `f` that maps from the logical layout to the physical layout, `f_inv(f(X))` can be simplified to `X` at the graph level, but `f(f_inv(X))` cannot.  Instead of simplifying into a memcpy, the transformations result in the following:

```python
@T.prim_func
def fused_inv_transform_Y_transform_Y(
    Y_write_cache: T.Buffer[(3, 8), "float32"],
    Y_read_cache: T.Buffer[(3, 8), "float32"],
):
    for io, ii in T.grid(3, 8):
        i = 8 * io + ii - 2
        if 0 <= i < 18:
            Y_read_cache[io, ii] = Y_write_cache[io, ii]
        else:
            Y_read_cache[io, ii] = 0.0
```

This expression could be simplified by using the original constraint on `Y_write_cache` provided in the layout transformation, but reconstructing that constraint couldn't be done by local analysis of any single function.

Reply:
@Lunderberg That's true. If the reduction dimension is padded, we will need to insert hint in the graph to assert it was previously padded by 0.

Reply:
Of course for the sake of discussion the example is limited to two convolutions. But the general case of multiple (N) back-to-back contractions with padded transformations, handling at the graph-level requires similar non-local information/hints across the sequence of operators to Nth order.

Reply:
Thanks for the discussions so far :) I think we all agree that having additional information variant in the buffer interface would results powerful expressiveness. Just like in high-level language while loop and for resulted in being more powerful.

Just like stated in the last statement. We could indeed consider put some constraints at the interface level. That would need some more thoughts on semantics, how would they interact with graph, and structural complexity (whether padding is something that worth the IR complexity).

On the other hand, I would still encourage us to think whether such complexity is worthwhile for only for padding. As a non-local back to back transformation padding is relatively easy and still achieved the goal.